elftools/dwarf/dwarfinfo.py

```diff
@@ -456,6 +456,34 @@ def _parse_line_program_at_offset(self, debug_line_offset, structs):
             self.debug_line_sec.stream,
             debug_line_offset)
 
+        # DWARF5: resolve names
+        def resolve_strings(self, lineprog_header, format_field, data_field):
+            if lineprog_header.get(format_field, False):
+                data = lineprog_header[data_field]
+                for field in lineprog_header[format_field]:
+                    def replace_value(data, content_type, replacer):
+                        for entry in data:
+                            entry[content_type] = replacer(entry[content_type])
+
+                    if field.form == 'DW_FORM_line_strp':
+                        replace_value(data, field.content_type, self.get_string_from_linetable)
+                    elif field.form == 'DW_FORM_strp':
+                        replace_value(data, field.content_type, self.get_string_from_table)
+                    elif field.form in ('DW_FORM_strp_sup', 'DW_FORM_strx', 'DW_FORM_strx1', 'DW_FORM_strx2', 'DW_FORM_strx3', 'DW_FORM_strx4'):
+                        raise NotImplementedError()
+
+        resolve_strings(self, lineprog_header, 'directory_entry_format', 'directories')
+        resolve_strings(self, lineprog_header, 'file_name_entry_format', 'file_names')
+
+        # DWARF5: provide compatible file/directory name arrays for legacy lineprogram consumers
```

Comment:
Here and elsewhere, could you please clarify what "legacy" means?

Reply:
On this one, I could use some advice. In lineprogram header v4 and below, there are two arrays - directories, which is an array of strings, and files, which is an array of (name, directory_index, last_mod_time, size) tuples. These are stored in the lineprogram header object as two collections - `include_directory` and `file_entry`, respectively.

In v5, both arrays are present but they have a dynamic structure - they are each prefixed with an array of fields. So I could not reuse the old variable names, `include_directory` and `file_entry`, without either losing data or breaking item datatype. I've introduced two new data items in the header - `directories` and `file_names`. The names came straight from the DWARF5 spec. However, since the v5 data structures can be (but don't have to be) supersets of the v4 ones, I've provided code that would generate `include_directory` from `directories` and `file_entry` from `file_names` if the latter are present. This way, if an existing library consumer encounters a binary with lineprogram header v5, they will still see the familiar data structures.

This changes the externally visible API, but not in a breaking way. Let me know if I should have handled this differently. The fact that file names and directory names are data items (as opposed to, say, method calls) is a given.

Reply:
I understand, thanks for clarifying. I think this all sounds fair and you should document it in a comment in the code.

Reply:
Commented in the `DWARFInfo.line_program_for_CU()` heredoc.